larq/layers_base.py

```diff
@@ -79,6 +88,79 @@ def get_config(self):
         }
 
 
+class QuantizerBaseConv(tf.keras.layers.Layer):
+    """Base class for defining quantized conv layers"""
+
+    def __init__(self, *args, pad_values=0.0, **kwargs):
+        self.pad_values = pad_values
+        super().__init__(*args, **kwargs)
+
+    def _is_native_padding(self):
+        return self.padding != "same" or (
+            not tf.is_tensor(self.pad_values) and self.pad_values == 0.0
+        )
+
+    def _get_spatial_padding_same(self, shape):
+        return tuple(
+            _compute_padding(stride, dilation_rate, input_size, filter_size)
+            for stride, dilation_rate, input_size, filter_size in zip(
+                self.strides, self.dilation_rate, shape, self.kernel_size
+            )
+        )
+
+    def _get_spatial_padding_same_shape(self, shape):
+        return [
+            size + sum(pad)
+            for size, pad in zip(shape, self._get_spatial_padding_same(shape))
+        ]
+
+    def _get_spatial_shape(self, input_shape):
+        return (
+            input_shape[1:-1]
+            if self.data_format == "channels_last"
+            else input_shape[2:]
+        )
+
+    def _get_padding_same(self, inputs):
+        padding = self._get_spatial_padding_same(self._get_spatial_shape(inputs.shape))
+        return (
+            ((0, 0), *padding, (0, 0))
+            if self.data_format == "channels_last"
+            else ((0, 0), (0, 0), *padding)
+        )
+
+    def _get_padding_same_shape(self, input_shape):
+        spatial_shape = self._get_spatial_padding_same_shape(
+            self._get_spatial_shape(input_shape)
+        )
+        if self.data_format == "channels_last":
+            return tf.TensorShape([input_shape[0], *spatial_shape, input_shape[-1]])
+        return tf.TensorShape([*input_shape[:2], *spatial_shape])
+
+    def build(self, input_shape):
+        if self._is_native_padding():
+            super().build(input_shape)
+        else:
+            with utils.patch_object(self, "padding", "valid"):
+                super().build(self._get_padding_same_shape(input_shape))
+
+    def call(self, inputs):
+        if self._is_native_padding():
+            return super(QuantizerBaseConv, self).call(inputs)
```

Comment:
Unfortunately this Python 2 syntax is needed to make autograph in TF 1.14 happy.

Reply:
6b80b1d had the nice side-effect that we can get rid of this workaround 🎉